contents/java/JUnit.md

```diff
@@ -0,0 +1,226 @@
+* `assertTrue`/`assertFalse`: checks whether statements return true/false as expected.
+* `assertNull`/`assertNotNull`: checks whether something is null or not, without causing a NullPointerException.
+* `assertArrayEquals`: yes, it loops through an entire array for us!
+* `assertThrows`/`assertDoesNotThrow`: checks that a piece of code throws the correct type of exception. In other words, JUnit even checks that the application can handle errors (or not)!
+* `assertTimeout`: ensures that a piece of code runs within time constraints - we don't want to keep the user waiting!
+
+These powerful expressions means that we can write more expressive tests.
+Instead of
+```java
+try {
+  int result = divide(100, 0); // Divides 100 by 0
+} catch (ArithmeticException ae) {
+  // This is the expected behaviour.
+  return;
+}
+// If code is here, then the ArithmeticException wasn't thrown.
+throw new AssertionError("Dividing 100 by 0 should throw an exception!");
+```
+
+We can now write
+```java
+assertThrows(ArithmeticException.class, () -> divide(100, 0));
+```
+
+This expression is a lot easier to understand, and as an added bonus, the code is also much shorter to write!
+
+### Before/After hooks
+
+Not every system can be tested so easily. Sometimes, testing is not as straightforward as running a function.
+For example, we might want to store information in a database. We would have to first connect to the database before we can run any test.
+
+JUnit exposes an API that allows developers to do this process easily and clearly. It offers the `@BeforeAll` and `@BeforeEach` tags that allow code to be run before each test, allowing us to easily initialise our testing environments before we run any tests:
+
+```java
+class TestInvolvingDatabases {
+  Database db;
+  @BeforeAll
+  public static void initializeDatabaseConnection() {
+    // Initialize database connections
+    db = connectToDatabase();
+  }
+  @Test
+  public static void addUserTest() {
+    db.addUser('NewUser');
+    ...
+  }
+}
+```
+
+The `BeforeEach` API can also be leveraged to ensure that we standardise the testing environment before each test.
+For example, we can clear our database before each test to ensure that tests do not <trigger for="modal:affectEachOther" trigger="click">affect each other</trigger>.
+
+```java
+class TestInvolvingDatabase {
+  Database db;
+  @BeforeEach
+  public static void resetDatabaseTable() {
+    db.truncate("testTable");
+  }
+}
+```
+
+### Extension Model
+
+Consider a project where multiple functions in different files each need to connect to a database with a different URL.
+One way would be to write a separate `initializeDatabaseConnection()` method for each test that requires a database connection. However, this violates the <popover effect="fade" content="Don't Repeat Yourself" placement="top">DRY</popover> principle.
+
+If our project is written in a OOP fashion that encourages abstraction and reuse, why shouldn't our tests be so too?
```

Comment:
The two things are not necessarily connected. Also, it drags in OOP into the picture without a good reason.